src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/QualityGuidelines/AvoidMultipleEnumerations/AvoidMultipleEnumerations.DataFlowOperationVisitor.cs

```diff
@@ -0,0 +1,146 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the MIT license.  See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+using System.Diagnostics.CodeAnalysis;
+using Analyzer.Utilities;
+using Analyzer.Utilities.FlowAnalysis.Analysis.GlobalFlowStateDictionaryAnalysis;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Operations;
+
+namespace Microsoft.CodeQuality.Analyzers.QualityGuidelines.AvoidMultipleEnumerations
+{
+    public partial class AvoidMultipleEnumerations
+    {
+        internal abstract class AvoidMultipleEnumerationsFlowStateDictionaryFlowOperationVisitor : GlobalFlowStateDictionaryFlowOperationVisitor
+        {
+            private readonly WellKnownSymbolsInfo _wellKnownSymbolsInfo;
+            private readonly IMethodSymbol? _getEnumeratorMethod;
+
+            protected AvoidMultipleEnumerationsFlowStateDictionaryFlowOperationVisitor(
+                GlobalFlowStateDictionaryAnalysisContext analysisContext,
+                WellKnownSymbolsInfo wellKnownSymbolsInfo,
+                IMethodSymbol? getEnumeratorMethod) : base(analysisContext)
+            {
+                _wellKnownSymbolsInfo = wellKnownSymbolsInfo;
+                _getEnumeratorMethod = getEnumeratorMethod;
+            }
+
+            protected abstract bool IsExpressionOfForEachStatement(SyntaxNode node);
+
+            public override GlobalFlowStateDictionaryAnalysisValue VisitParameterReference(IParameterReferenceOperation operation, object? argument)
+            {
+                var value = base.VisitParameterReference(operation, argument);
+                if (!IsDeferredType(operation.Parameter.Type.OriginalDefinition, _wellKnownSymbolsInfo.AdditionalDeferredTypes))
+                {
+                    return value;
+                }
+
+                if (!IsOperationEnumeratedByMethodInvocation(operation, _wellKnownSymbolsInfo) && !IsGetEnumeratorOfForEachLoopInvoked(operation))
+                {
+                    return value;
+                }
+
+                if (!TryGetInvocationEntity(operation, out var invocationEntity))
+                {
+                    return value;
+                }
+
+                var newValue = CreateAnalysisValue(invocationEntity, operation, value);
+                UpdateGlobalValue(newValue);
+                return newValue;
```

Comment:
Can we move this code into a shared private helper method and avoid duplicating it below in VisitLocalReference?